tb_plugin/torch_tb_profiler/profiler/overall_parser.py

```diff
@@ -177,22 +177,99 @@ def __init__(self):
+    # Update self.steps considering device side events launched by each host side step.
+    def update_steps_consider_device_side(self, runtime_node_list, device_node_list):
+        runtime_node_list = sorted(runtime_node_list, key=lambda x: x.start_time)
+        # Make sure self.steps is sorted by time.
+        self.steps = sorted(self.steps, key=lambda x: x[0])
+        # Use similar code with two-way merge to get all runtimes inside each host-side step span,
+        # then record each step's min kernel start time and max kernel end time:
+        steps_device = [(sys.maxsize, -sys.maxsize - 1)] * len(self.steps)
+        i_step = 0
+        i_runtime = 0
+        step_device_min_ts = sys.maxsize
+        step_device_max_ts = -sys.maxsize - 1
+        matched_device_nodes = set()
+        while i_step < len(self.steps) and i_runtime < len(runtime_node_list):
+            step_host_start_time = self.steps[i_step][0]
+            step_host_end_time = self.steps[i_step][1]
+            if runtime_node_list[i_runtime].start_time < step_host_start_time:
+                # This runtime is ahead of or intersects with this step span. Skip this runtime.
+                i_runtime += 1
+            elif runtime_node_list[i_runtime].end_time <= step_host_end_time:
+                # and runtime_node_list[i_runtime].start_time >= step_host_start_time
+                # This runtime is inside this step span. Scan its device_nodes.
+                rt = runtime_node_list[i_runtime]
+                if rt.device_nodes is not None:
+                    for device_node in rt.device_nodes:
+                        step_device_min_ts = min(device_node.start_time, step_device_min_ts)
+                        step_device_max_ts = max(device_node.end_time, step_device_max_ts)
+                        matched_device_nodes.add(device_node)
+                i_runtime += 1
+            elif runtime_node_list[i_runtime].start_time < step_host_end_time:
+                # and runtime_node_list[i_runtime].end_time > step_host_end_time
+                # This runtime intersects with this step span. Skip this runtime.
+                i_runtime += 1
+            else:
+                # runtime_node_list[i_runtime].start_time >= step_host_end_time
+                # This runtime starts after this step's end. Record and move forward this step.
+                steps_device[i_step] = (step_device_min_ts, step_device_max_ts)
+                i_step += 1
+                step_device_min_ts = sys.maxsize
+                step_device_max_ts = -sys.maxsize - 1
+        while i_step < len(self.steps):
+            # This step doesn't launch any device side event, just assign it as empty.
+            steps_device[i_step] = (step_device_min_ts, step_device_max_ts)
+            i_step += 1
+        # Change step time to device side on the condition that any step have device time.
+        is_use_gpu = False
+        for steps_device_item in steps_device:
```

Comment:
replace the for loop with 'matched_device_nodes is not empty' ?

Reply:
Thanks. Done.